### PR TITLE
Feature/device integration testing

### PR DESCRIPTION
**Pull request by bporter125**
Still a little rough, but I'd like to get some eyes on it. It's overall pretty fleshed out, and all that I have left is to maybe add a few more scenarios as well as fix some instability issues (Appium crashing sometimes). Also note that Kobiton has not fully passed the suite as there are issues with animation timings that still need to be addressed.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -5,6 +5,7 @@ verify_ssl = true
 
 [packages]
 launchkey = {path = ".", editable = true}
+appium-python-client = "*"
```
Comment (lordkyzr) [inline Pipfile]: Should this be a dev dependency?

File: features/example_environment.py [hunk 2]
```diff
@@ -35,6 +35,22 @@ def before_all(context):
...
+
+    # Note that that the only app currently supported is the Android
+    # Authenticator SDK demo app
+    # https://github.com/iovation/launchkey-android-authenticator-sdk/tree/master/demo-app
+    context.sample_app_apk_path = '/path/to/sample-app.apk'
+
+    # You can use Kobiton Testing
+    context.kobiton_username = "username"
+    context.kobiton_sdk_key = "83aab95e-c577-410d-a7db-c7a83bd2433a"
+
```
Comment (aenglander) [inline features/example_environment.py]: You may want to mention that the Kobiton entries will have to be commented out or removed to use Appium.

File: features/service_client/auth_request/service-client-auth-request-get-device-response-methods.feature [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    And I approve the auth request
+    And I get the response for the Authorization request
+    Then the Authorization response should contain the following methods:
+      | Method      | Set   | Active | Allowed | Supported  | User Required | Passed | Error |
+      | wearables   | False | False  | True    | True       | Null          | Null   | Null  |
+      | geofencing  | Null  | True   | True    | True       | Null          | Null   | Null  |
+      | locations   | False | False  | True    | True       | Null          | Null   | Null  |
+      | pin_code    | False | False  | True    | True       | Null          | Null   | Null  |
+      | circle_code | False | False  | True    | True       | Null          | Null   | Null  |
+      | face        | False | False  | True    | False      | Null          | Null   | Null  |
```
Comment (aenglander) [inline features/service_client/auth_request/service-client-auth-request-get-device-response-methods.feature]: I'm a little worried about what happens when an Android device has supported face recognition. We either need to ensure face capabilities is not supported or document the feature saying that if a device supports facial recognition that the scenario will fail

File: features/service_client/auth_request/service-client-auth-request-get-device-response-policy.feature [hunk 4]
```diff
@@ -0,0 +1,37 @@
+Feature: Service Client Authorization Request: Get Device Response Policy
```
Comment (aenglander) [inline features/service_client/auth_request/service-client-auth-request-get-device-response-policy.feature]: We probably need a check for named geofences. That would have to be set in the Service policy I believe.

File: features/steps/managers/kobiton_manager.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
+import requests
+from time import sleep
+
+
+class Version:
+    def __init__(self, id, state=None, version=None, native_properties=None, latest=None):
```
Comment (aenglander) [inline features/steps/managers/kobiton_manager.py]: Can we document the parameters for clarity

File: features/steps/managers/kobiton_manager.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        self.id = id
+        self.state = state
+        self.version = version
+        self.native_properties = native_properties
+        self.latest = latest
+
+
+class App:
+    def __init__(self, id, name=None, state=None, created_at=None, private_access=None, os=None, created_by=None,
```
Comment (aenglander) [inline features/steps/managers/kobiton_manager.py]: Can we document the parameters for clarity

File: features/steps/managers/kobiton_manager.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    def __repr__(self):
+        return "App <id={id}, name=\"{name}\", state=\"{state}\">".format(
+            id=self.id,
+            name=self.name,
+            state=self.state
+        )
+
+
+class Device:
+    def __init__(self, id, udid, is_booked, is_hidden, is_online, model_name, device_name,
```
Comment (aenglander) [inline features/steps/managers/kobiton_manager.py]: Can we document the parameters for clarity

File: features/steps/managers/kobiton_manager.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        self.is_my_org = is_my_org
+        self.is_my_own = is_my_own
+        self.hosted_by = hosted_by
+
+    def __repr__(self):
+        return "Device <{device_name}>".format(device_name=self.device_name)
+
+
+class KobitonManager:
+    def __init__(self, username, sdk_key, url='https://api.kobiton.com', api_version='v1'):
```
Comment (aenglander) [inline features/steps/managers/kobiton_manager.py]: Can we document the parameters for clarity

File: features/steps/managers/appium_device_manager.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
+from appium import webdriver
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.ui import WebDriverWait
+from selenium.webdriver.support.wait import TimeoutException
+from appium.webdriver.webdriver import MobileBy
+
+
+class AppiumDeviceManager:
+
+    def __init__(self, appium_url, device_capabilities, app_package,
```
Comment (aenglander) [inline features/steps/managers/appium_device_manager.py]: Can we document the arguments for clarity?

**Comment by aenglander** [comment]
It looks like were are doing a lot of manual assertions that may be resolved by PyHamcrest. It would probably be a better way to make those assertions.

**Comment by aenglander** [comment]
Could we document the public methods of the new classes? Without them, we require a knowledge of Kobiton and Appium that probably isn't necessary.

File: features/steps/managers/kobiton_manager.py (hunk 7 above)
Comment (bporter125) [reply features/steps/managers/kobiton_manager.py]: Honestly no, and this goes for most Kobiton objects. They have near to no documentation regarding return values, so these parameters are based on what I saw being returned. I'll add a Pydoc but documenting each of the values won't be beneficial.

File: features/example_environment.py (hunk 2 above)
Comment (bporter125) [reply features/example_environment.py]: It's the opposite. But I'll document that.

File: features/steps/managers/appium_device_manager.py (hunk 9 above)
Comment (bporter125) [reply features/steps/managers/appium_device_manager.py]: Yep, I'll also document the methods.

File: features/steps/managers/kobiton_manager.py (hunk 8 above)
Comment (bporter125) [reply features/steps/managers/kobiton_manager.py]: Yep, I'll also document the methods.

File: features/service_client/auth_request/service-client-auth-request-get-device-response-policy.feature (hunk 4 above)
Comment (bporter125) [reply features/service_client/auth_request/service-client-auth-request-get-device-response-policy.feature]: I'll add a second scenario to test named geofences so we can test both.

File: features/base_environment.py [hunk 10]
```diff
@@ -14,10 +15,68 @@ def before_all(context):
...
+            timeout_period=10
+        )
+    elif hasattr(context, 'kobiton_username') and hasattr(context,
+                                                          'kobiton_sdk_key'):
+        context.kobiton_manager = KobitonManager(context.kobiton_username,
+                                                 context.kobiton_sdk_key)
+        context.uploaded_app = context.kobiton_manager.upload_app(
+            context.sample_app_apk_path)
+
+        devices = context.kobiton_manager.get_devices()
```
[the remark this answers is not in the document]
Comment (jnpblrdrgz) [reply features/base_environment.py]: I re-read  this all well and without some documentation it's just confusing.

File: features/steps/managers/appium_device_manager.py (hunk 9 above)
Comment (bporter125) [reply features/steps/managers/appium_device_manager.py]: Done

File: features/steps/managers/kobiton_manager.py (hunk 8 above)
Comment (bporter125) [reply features/steps/managers/kobiton_manager.py]: Done

File: features/steps/managers/kobiton_manager.py (hunk 7 above)
Comment (bporter125) [reply features/steps/managers/kobiton_manager.py]: Added PyDoc and link to the Kobiton documentation (which is as bad as I suggested).

File: features/steps/managers/kobiton_manager.py (hunk 6 above)
Comment (bporter125) [reply features/steps/managers/kobiton_manager.py]: Added PyDoc and link to the Kobiton documentation

File: features/steps/managers/kobiton_manager.py (hunk 5 above)
Comment (bporter125) [reply features/steps/managers/kobiton_manager.py]: Added PyDoc and link to the Kobiton documentation

File: features/service_client/auth_request/service-client-auth-request-get-device-response-policy.feature (hunk 4 above)
Comment (bporter125) [reply features/service_client/auth_request/service-client-auth-request-get-device-response-policy.feature]: Added

File: features/example_environment.py (hunk 2 above)
Comment (bporter125) [reply features/example_environment.py]: Done

File: features/base_environment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (bporter125) [reply features/base_environment.py]: Added documentation stating what it was filtering and why. Also added an error condition for the case that it doesn't find a device that works.

File: Pipfile (hunk 1 above)
Comment (bporter125) [reply Pipfile]: Good find. Updated.

File: features/service_client/auth_request/service-client-auth-request-get-device-response-methods.feature (hunk 3 above)
Comment (bporter125) [reply features/service_client/auth_request/service-client-auth-request-get-device-response-methods.feature]: To be addressed in a future PR.

**Pull request merged**